src/urllib3/connectionpool.py

```diff
@@ -938,10 +940,15 @@ def _prepare_conn(self, conn):
 
     def _prepare_proxy(self, conn):
         """
-        Establish tunnel connection early, because otherwise httplib
-        would improperly set Host: header to proxy's IP:port.
+        Establishes a tunnel connection through HTTP CONNECT.
+
+        Tunnel connection is established early because otherwise httplib would
+        improperly set Host: header to proxy's IP:port.
         """
-        conn.set_tunnel(self._proxy_host, self.port, self.proxy_headers)
+
+        if self.proxy.scheme != "https":
+            conn.set_tunnel(self._proxy_host, self.port, self.proxy_headers)
```

Comment:
You're losing the `proxy_headers` on the `conn.connect()`

Reply:
Thanks @aausch, great catch.

I've added another commit in which we are now adding the proxy headers when using an HTTPS proxy. Instead of sending them with the HTTP CONNECT, we'll be including them in the requests forwarded to the proxy similar to what's done for HTTP requests.

Reply:
@jalopezsilva merging the headers is incorrect behavior. Unless I'm missing changes elsewhere, it looks to me like there are multiple problems with the commit as is:

1. No tunnel is actually being established. As far as I can tell, the `CONNECT` message is never being sent. When the `CONNECT` message does go out, it's expected to contain the authentication portion, at the very least, for the proxy headers. Though it probably will need all the headers since that's the last message that will be sent to the proxy without the inner connection's encryption layer?
2. The headers should not be merged. Proxy headers only should be sent to the proxy, on the outer-tunnel connection, and not merged then sent on a mixed connection or inner connection (depending on how you think of the two SSL encryption layers)
3. The inner connection is not getting wrapped in an SSL wrapper at any point; this will prevent any attempts to connect to HTTPS sites through the proxy. Also, this is a non trivial exercise, since urllib, and the underlying python C wrappers around the OpenSSL, don't provide native wrappers. And the OpenSSL wrapper that's passed all the way up doesn't support wrapping encrypted connections/sockets - it only supports OS native sockets. See for example the conversation here: https://bugs.python.org/issue29394

Reply:
see https://github.com/python/cpython/blob/3.8/Lib/http/client.py#L883 ->

1. I'm under the impression httplib doesn't appear to properly support encrypted tunnels as of Python 3.8 (and still only connects to proxies via `HTTP 1.0` instead of 1.1!)
2. I'm fairly confident (unless I'm missing something in one of the libraries?) that without an explicit call to `_tunnel`, you won't end up establishing a proxy connection

Reply:
Again, maybe I'm missing something and this code just works? But what I expected to see isn't this CR

---

For `http` proxies, the existing code works - it can just reuse the connection into the proxy:
1. get a new connection, to the proxy
2. send the transparent connect/ tunnelling message request
3. wrap the proxy connection in SSL encryption aimed at the `https` site we want to reach (which bottoms out in calls to OpenSSL or w/e on the underlying system, and properly upgrades the connection/socket/file descriptor)
4. shoot encrypted messages to the proxy and they'll get bounced through

---

This won't work for https proxies; what has to happen:
1. get a new (http) connection to the proxy
2. escalate to/ wrap w the proxy's `https` certificates
3. sent a `CONNECT` message/open a tunnel
4. create an in-memory equivalent of a buffered connection/socket/fd, plugging the output into the proxy's https encrypted connection/socket/fd
5. wrap the in-memory buffered connection/socket/fd with the end sites' `https` certificate
6. use the in memory buffered connection/socket/fd for any further communication

---

Currently, I don't think step 3 is happening anywhere. I would expect that steps 4-6 should really be happening somewhere inside of the OpenSSL library, or, alternatively, somewhere in the CPython implementation, for reasons of CPU performance and to avoid memory pressure issues, though the https://bugs.python.org/issue29394 thread seems to suggests a Python level alternative is available? At any rate, I don't see those steps addressed in this CR, but maybe I'm missing a change in a library upgrade or elsewhere in middleware

Reply:
@aausch,  Thanks for the review. Your understanding of the PR is correct, I am intentionally not establishing the tunnel through HTTP CONNECT. I am instead establishing a TLS connection to the proxy and forwarding requests to the proxy within that connection. The proxy should establish the TLS connection with the destination and forward the requests upstream. I tried to be clear that was my intent in my first comment of the PR (apologies if it wasn't clear enough!)

The implementation that I chose avoids the client having to do TLS within TLS. The connection to the destination is secure: there's a TLS connection between client<->proxy and proxy<->server. The only disadvantage is that the client won't have a direct TLS session with the destination. (Which prevents mutual authentication with the destination) There's also an implicit trust set on the proxy by the client. Given that we're explicitly trusting the proxy by establishing a TLS session to it I thought this was agreeable. Support in forward proxies for this feature is relatively common (although we can do a quick audit and verify).

The other implementation alternative would be to perform TLS to the proxy, HTTP CONNECT to the destination and TLS to the destination within the tunnel. This lines up with what you outlined in the comment above. As you said, implementing this is not trivial. It is indeed possible within Python if you use the ``SSLContext.wrap_bio`` methods. The tricky part of implementing it through the ``wrap_bio`` methods is that you have to manage the IO within the socket and the memoryBIO yourself. You can see an example of how this looks like on the following ``CPython`` test case: https://github.com/python/cpython/blob/master/Lib/test/test_ssl.py#L2150-L2235.  I have a tentative implementation of an ``SSLTransport`` object which takes a socket, wraps it and performs the IO within itself using the ``wrap_bio`` APIs. It would work like an ``SSLSocket`` except it wouldn't take ownership of the underlying native socket. 

I think we have three options for moving forward:

1. **Leave this PR as is and merge it** - The PR will add initial support for HTTPS proxies. We can add documentation indicating what the behavior is and the limitations.

2. **Add TLS within TLS support to urllib3** - I can add the ``SSLTransport`` object I mentioned before to urllib3/contrib/ssl.py and only use it when we need to do TLS within TLS. It will probably take us a couple of iterations to get it 100% right but we can give it a shot. 

3. **Add TLS within TLS within CPython** - Similar to #2 but I try to upstream the ``SSLTransport`` object to CPython. This will probably take the longest time as we would have to wait for Python 3.9 or even beyond.

I'm happy to move forward with any of the options. Out of the three, #3 is probably my least favorite given that it would take a long time to get this feature out. Ultimately I think we need an easy way to do TLS within TLS in Python. If we go ahead with #2, I could try doing #3 once we have a stable implementation within urllib3.

Reply:
>The implementation that I chose avoids the client having to do TLS within TLS. The connection to the destination is secure: there's a TLS connection between client<->proxy and proxy<->server. 

yes, I saw that this is your intent. But the code as is does not achieve this, as far as I can tell.  'tunneling' means 
1. 'tell the proxy to create the second connection' ie. what you call `proxy <-> server` in your comment, and also
2. tell the proxy to blindly forward messages along it to a specific server. 

the specific protocol for it, is to send a special version of the http connect message - something like  `CONNECT <ultimate destination server>` - to the proxy. Without that message, the proxy doesn't know where to forward messages to/what connection to open on our behalf.  as far as I can tell, this pull request doesn't, anywhere, tell the proxy to open the second connection and to which server to open it to.

>  The only disadvantage is that the client won't have a direct TLS session with the destination.

That is a significant disadvantage. Would you be ok if your proxy provider can read, for example, your authentication information for logging in to your bank? or your login information for Netflix?  Your keys for making a bitcoin exchange? Assuming you can deliver on this, you're guaranteeing a MITM will read all of that information, and do whatever they want with it. 

I'm not sure if there is a legitimate use case you're looking to address here (maybe developers who are using a mock proxy during code testing?), but, if and assuming there is one, I would:

1. guard this specific use case a flag (you'd want the end user to explicitly set a flag "allow proxy to read all my information" to true or otherwise indicate they're interested in trusting the proxy with their info, so maybe add an argument into the function and check it)
2. spit out a large Warning message in bright, bold, red text, across every possible readable output anytime anything is sent across this connection

>  Given that we're explicitly trusting the proxy by establishing a TLS session to it I thought this was agreeable.

This is a **significant** error or misinterpretation of what TLS is, and what it does. TLS **does not imply any kind of trust**; it just protects you from having your ISP, or a government, or law enforcement, or your friendly hacker on a Starbucks wifi connection, know that you're using a proxy. 

Also, per my comment above and more generally, even if there was a method for trusting proxies in general, I would highly recommend against trusting proxies with most of the things that you would use a proxy for.

> Support in forward proxies for this feature is relatively common (although we can do a quick audit and verify).

If by "this feature" you mean anything other than tunnelling, then I'm really not sure which feature you're talking about - please provide a reference or more description of it.

---

> 1. **Leave this PR as is and merge it** 

As far as I can tell, this PR as is does not actually provide any additional value, and should not yet be merged. It does allow opening of HTTPS connections to proxies, which is an improvement, but then does not go on to create a tunnel or, through any other protocol, tell the proxy it should pass messages on to a specific target server - so no messages are then sent past the proxy to the ultimate target server. Also, it introduces two significant bugs, caused by the merging of the headers:
- as written, all the headers are only being sent into the proxy on the initial connect into it, and won't be passed through to the ultimate target server; also, i believe the headers are no longer correctly supporting authentication into the proxy, so if you're dealing with a proxy that expects you to log in to it, the connect will fail
- i'm not sure if the proxy will know how to handle the merged headers at all. think about it this way: what if the second server you're connecting to, is also a proxy, which is expecting certain proxy headers? how will the first proxy know which proxy headers it should interpret, and which proxy headers it should send along to the second server?

The other two options are, I think, the only options which will allow proxying into an endserver via the HTTPS protocol, and one of them should eventually be implemented (also assuming the bugs listed above are fixed). I also think there's a possible intermediate option:

1.5: **Fix this PR so that it at least works correctly for the  `TLS <> proxy`, `proxy <> HTTP site` use case, then merge it**

I believe this PR is pretty close, as is, to supporting forwards to HTTP sites only, over a TLS proxy connection. To get there, just make sure that after the connection is tls wrapped, you send the correct command to establish a tunnel to the end server(including the proxy headers and correctly assembled proxy authentication information). 

As a bonus, it would be also awesome to switch to the `HTTP 1.1 ` version of the tunnelling request, since that allows for persistent connections through the proxy and I suspect would be much more efficient than reopening the connection every time

Reply:
Oh, and, if going with option `1.5`, probably check if the end site protocol is tls via a tls proxy, and if it is, spit out a warning or error warning the user that it's not yet supported, but it will be supported soon

Reply:
> As far as I can tell, this PR as is does not actually provide any additional value, and should not yet be merged. It does allow opening of HTTPS connections to proxies, which is an improvement, but then does not go on to create a tunnel or, through any other protocol, tell the proxy it should pass messages on to a specific target server - so no messages are then sent past the proxy to the ultimate target server.

@aausch, I'm a little surprised by this comment and I want to make sure we are in the same page. 

When proxies are presented with a absolute URI in a request, ``GET http://example.com``, the proxy will forward that request for you. This is the expected behavior when providing an absolute URI to the proxy. (Documented [here](https://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html)).  The proxy will establish a TCP connection and forward the request to the destination.

urllib3 uses this when requesting HTTP resources through an HTTP proxy. Notice how in ``HTTPConnectionPool``, ``_prepare_proxy`` doesn't establish any tunnel. The request is simply forwarded with the absolute URI. (See lines 327-330, ``poolmanager.py``). There's no CONNECT request made by urllib3 to the proxy. When fetching HTTPS resources through an HTTP proxy we do send the CONNECT as we need to establish the TLS session to the destination. (``HTTPSConnectonPool``, ``_prepare_proxy`` does this). Here's an excerpt of a tcpdump verifying that: https://pastebin.com/ZNyBWH5B 

This PR re-uses the forwarding functionality to fetch resources from an HTTPS proxy. I agree that fetching HTTPS resources from an HTTPS proxy is insecure and I'll be adding CONNECT support for that, but fetching HTTP resources through an HTTPS proxy works as intended in this PR. See the integration tests that I added with (test/with_dummyserver/test_proxy_poolmanager.py) which verify that.  I have also added this local patch to integration tests in my company as well. 

> * as written, all the headers are only being sent into the proxy on the initial connect into it, and won't be passed through to the ultimate target server; also, i believe the headers are no longer correctly supporting authentication into the proxy, so if you're dealing with a proxy that expects you to log in to it, the connect will fail
> * i'm not sure if the proxy will know how to handle the merged headers at all. think about it this way: what if the second server you're connecting to, is also a proxy, which is expecting certain proxy headers? how will the first proxy know which proxy headers it should interpret, and which proxy headers it should send along to the second server?

The header behavior is the same that urllib3 uses when fetching HTTP resources with an HTTP proxy. In that case headers are also merged. (See src/urllib3/connectionpool.py, line 637). This works relatively well. The proxy will remove [per-hop headers](https://tools.ietf.org/html/rfc2616#section-13.5.1) to prevent the server from seeing headers it shouldn't see and forward the request.

The case which isn't supported is a second proxy which might expect a second set of headers. This currently doesn't work on the HTTP proxy / HTTP resource case. I think we can continue merging headers on the HTTPS proxy / HTTP resource case as well.


## Moving forward
Let me know if this makes sense. 

I'll throw an exception when users attempt to fetch HTTPS resources through an HTTPS proxy as that's insecure as we agreed until I add the TLS within TLS support.. I can also add a parameter ``allow_insecure_proxy_to_see_my_traffic`` if we want to keep that behavior configurable by the client.

Reply:
> When proxies are presented with an absolute URI in a request, GET http://example.com, the proxy will forward that request for you. This is the expected behavior when providing an absolute URI to the proxy. (Documented here). The proxy will establish a TCP connection and forward the request to the destination.

Yeah, you're right! I forgot about this usage pattern - I typically think of it as more of a bridge, than a proxy, and I forget that historically many proxies were nothing but bridges. Without a tunnel, the client has zero control over what happens (and fully exposes the communication to the proxy),  and most of the internet has moved past the point where bridges work correctly, at all, for most of what everyone does online.

>  I don't see the value of doing an HTTP CONNECT to fetch a HTTP resource

Without a tunnel, you only have what I think of as a bridged connection past the proxy. In general, that severely limits what you can do when communicating with the downstream server (there are lots of ways to hack more value out of it, but they're one offs and hacky) - eg., you can't hop over multiple proxies, open sockets/websockets, perform correct https/tls, play most modern video games, and most other web communication doesn't work correctly, etc... - i.e., you're supporting at most 20% of the kind of traffic we see on the internet, but probably closer to 10%. 

[aside: I am surprised, honestly, to be reminded that this kind of traffic is something people still use for anything, beyond highly specialized use cases (eg., when behind a corporate wall). It's also runs the risk of being horrendously inefficient! A tunnel, esp. over HTTP 1.1, has stronger guarantees to be a persistent, reusable, connection, which significantly speeds everything along.]

---

> This PR re-uses the forwarding functionality to fetch resources from an HTTPS proxy. I agree that fetching HTTPS resources from an HTTPS proxy is insecure and I'll be adding CONNECT support for that, but fetching HTTP resources through an HTTPS proxy works as intended in this PR. 

I agree that this is definitely something that needs to be supported and clearly useful/of value. I am surprised to find out that urllib3 does not yet support using tunnelling for anything but HTTPs traffic - but that is out of scope here and probably that should be a separate feature request (eg. dynamically figure out if tunnelling is available, default to a tunnel, and only if failing that). 

---

> Let me know if this makes sense. I don't see the value of doing an HTTP CONNECT to fetch a HTTP resource, unless you're concerned that the proxy defined as an "https proxy" in the client configuration proxy only supports CONNECT requests? Modern proxy servers support both though, CONNECT and forwarding requests.

I agree with almost everything. Definitely, HTTP CONNECT should be the default for all web communication, and basic forwarding should be just a fall back, but that work is out of scope I think for this PR and it makes sense for me to be a part of a different pull request.

> I'll throw an exception when users attempt to fetch HTTPS resources through an HTTPS proxy as that's insecure as we agreed until I add the TLS within TLS support.. I can also add a parameter allow_insecure_proxy_to_see_my_traffic (default False of course) if we want to keep that behavior configurable by the client.

If it's guarded by a False default flag, it's ok with me. Looking at the code, it would even make sense to me to just not support TLS, without tunnelling in place. It's a very niche, insecure, corner case, and wasn't supported historically by urllib3, probably due to the security concerns? I now wonder if there are (non-shady) proxies which support this, even.

Reply:
In total agreement @aausch, moving to HTTP CONNECT for everything instead of "bridging" is much more flexible but you're totally right that we probably should tackle this as a separate PR. Some people might be relying on the HTTP resource / HTTP proxy "bridging" behavior. 

I am now checking it we are attempting to fetch HTTPS resources through an HTTPS proxy and throwing an exception as this is not supported and insecure. I added a flag ``allow_insecure_proxy_to_see_my_traffic``, defaulted to False, and can be changed if a client wants the behavior.

Once we close this PR, I'll follow up with the TLS in TLS support to that we can close it out.